### ISSUE TITLE
[$250] [HelpDot] back should go to previous page

### ISSUE THREAD

**Issue by rushatgabhane** [Expensify/App#42600]
https://expensify.enterprise.slack.com/archives/C02QSAC6BJ8/p1716512720894059

1. When navigating on help.expensify.com, back should take you to the previous page
2. LHN highlight position isn't accurate 

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0147f1debc6f7ab719</li>
        <li>Upwork Job ID: 1801285227792706387</li>
        <li>Last Price Increase: 2024-08-20</li>
<li>Automatic offers: </li>
<ul>
<li>wildan-m | Contributor | 103733105</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @rushatgabhane</details>

**Comment by maddylewis** [comment]
asking for update here -- https://expensify.slack.com/archives/C02QSAC6BJ8/p1716906559944779

**Comment by rushatgabhane** [comment]
not overdue

**Comment by rushatgabhane** [comment]
im exploring browser apis to achieve this

**Comment by rushatgabhane** [comment]
still didn't find anything

**Comment by maddylewis** [comment]
not a priority rn as the changes require internal eng resources -- closing for now

**Comment by maddylewis** [comment]
this will go through the standard BZ process so reassigning mitch as the original BZ person 👍

**Comment by Taiwrash** [comment]
Hello,

I will love to jump on on this but I have the following clarification questions
- The back button described here, is it on mobile or web?
- Which page is affected, let me start with a triage and after I can solve it
- Pointing me to the right and giving a more descriptive issue description will really helps
Thanks, Looking forward to your help.
Best,
Rasheed

**Comment by Taiwrash** [comment]
<img width="1440" alt="Screenshot 2024-06-13 at 23 30 17" src="https://github.com/Expensify/App/assets/49725691/419212af-20eb-4b5b-91af-e80dca0aa9c4">
I want to check more on the stackoverflow link but i am getting this error

**Comment by MitchExpensify** [comment]
@Taiwrash 

- The back button described here, is it on mobile or web?

Both (See video)

- Which page is affected, let me start with a triage and after I can solve it Pointing me to the right and giving a more descriptive issue description will really helps

help.expensify.com/articles/new-expensify/chat/create-a-new-chat#start-a-chat-room

https://github.com/Expensify/App/assets/36425901/2be9c10b-a1ad-413f-ae86-5e81e2b15809

**Comment by Taiwrash** [comment]
Thanks for the explanation. I am on it...

**Comment by MitchExpensify** [comment]
How is progress on the proposal coming along @Taiwrash ?

**Comment by Taiwrash** [comment]
Thanks for this.

### Please re-state the problem that we are trying to solve in this issue.
- Navigating back from deep index should take you to the previous page and
- The LHN highlight position is inaccurate, the hover effects on the link/tabs are not correctly highlighting  when the back link tabs are clicked.

### What is the root cause of that problem?
The root course is browser based. see the demo on brave browser especially the highlight on LHN.

### What changes do you think we should make in order to solve the problem?
1. Create route back navigation by directing the back navigation to previous displayed page
2. Setup all complete browser compatibility by using CSS styles that work across all platforms and browser


https://github.com/Expensify/App/assets/49725691/c3de6d9a-1c45-4263-8d67-9e70a375eed0

**Comment by Taiwrash** [comment]
- See the test on Microsoft Egde browser below

Uploading Screen Recording 2024-06-18 at 10.06.44.mov…

**Comment by Taiwrash** [comment]
See the `Back` button template which handle the back action:

```html
       <!-- LHN content (Main navigation tree or Article TOC) -->
        <div id="lhn-content">
            {% if page.url == "/" or page.url contains "/hubs/" %}
                {% include lhn-template.html %}
            {% else %}
                <div id="back-button" class="icon-with-link">
                    <img src="/assets/images/back-left.svg" class="base-icon"></img>
                    <div class="link">Back</div>
                </div>
                <div class="article-toc"></div>
            {% endif %}
        </div>
```

**Comment by Taiwrash** [comment]
Yes!

Here is the [Demo](https://legendary-zebra-wgjw5wqjw4gf9xjg-4000.app.github.dev/) of the fix.

And here is the [PR URL](https://github.com/Expensify/App/pull/43910)

**Comment by rushatgabhane** [comment]
@Taiwrash we need the browser back button to behave in the similar manner as the LHN back button. Can you please integrate that into your proposal?

**Comment by Taiwrash** [comment]
> @Taiwrash we need the browser back button to behave in the similar manner as the LHN back button. Can you please integrate that into your proposal?

@rushatgabhane @MitchExpensify I just updated the proposal.

BTW, Did anyone check the [PR](https://github.com/Expensify/App/pull/43910)?

**Comment by Taiwrash** [comment]
@MitchExpensify @rushatgabhane @eVoloshchak still await final approval here

**Comment by Taiwrash** [comment]
I am still await your response here!

**Comment by Taiwrash** [comment]
Unfortunately I didn't get any notice here again.

CC: @s77rt  can you help check

**Comment by rushatgabhane** [comment]
> Handle the browser back button to behave in the similar manner as the LHN back button by changing the default ability of the browser

can you please explain how will this be achieved?

**Comment by Taiwrash** [comment]
> > Handle the browser back button to behave in the similar manner as the LHN back button by changing the default ability of the browser
> 
> can you please explain how will this be achieved?

- The `default.html` under `_layouts` folder under the `docs` page contain the code that handle the `back` LHN button routing. This is routing will be pointed to the previous page as requested.
- You can check this [PR](https://github.com/Expensify/App/pull/43910) where I route the `back` button to home as a way to test the right routing to previous page

**Comment by rushatgabhane** [comment]
> The default.html under _layouts folder under the docs page contain the code that handle the back LHN button routing. This routing will be pointed to the previous page as requested.

> You can check this https://github.com/Expensify/App/pull/43910 where I route the back button to home as a way to test the right routing to previous page

thank you! @Taiwrash i'd really appreciate it if you could explain the proposed changes in plain english. It's preferred over checking out a PR.


Also, we need the **browser** back button to behave in the similar manner as the LHN back button. How would that be achieved?

**Comment by Taiwrash** [comment]
Oh thanks for the clarification!

To set browser back button to behave as our custom `lhn-button`, we need to manage the browser's history and ensure that navigating back triggers the same actions as clicking the custom back button.

- We will use the `pushState` method to add entries to the browser's history when navigating within the app. With this, we can control the state of the history and handle custom actions when the user clicks the browser's back button.

- We need ensure that the same actions triggered by our custom back button are also triggered when the browser's back button is used. This can be done by handling the `popstate` event.

# My Code Implementation approach in `main.js`

```js
         function handleBackNavigation() {
            // Custom actions for back navigation: for example, navigate to the previous page or update UI elements
            window.history.back();
        }

        // Handle browser back button and custom actions
        window.addEventListener('popstate', function (event) {
            handleBackNavigation();
        });

        // Push initial state to history
        window.history.replaceState({ page: 'initial' }, '', window.location.href);

        // Function to push state to history (call this function when navigating within your app)
        function pushStateToHistory(state, url) {
            window.history.pushState(state, '', url);
        }
```

**Comment by rushatgabhane** [comment]
@Taiwrash in the demo you've shown, the back button scrolls up the page (navigates in history). 
But it should take you to the hub page

**Comment by Taiwrash** [comment]
Now I get the point, what I understood before now was just to make the browser back button behave save way the LHN back button behaves but now  it is now clear to me to make back button works by taking users to the hub page:

## Example
- If the current URL is `https://help.expensify.com/articles/new-expensify/getting-started/Create-a-company-workspace`, the script will change it to `https://help.expensify.com/new-expensify/hubs/getting-started/`
- If the current URL is `https://help.expensify.com/articles/new-expensify/some-other-section/another-topic`, the script will change it to `https://help.expensify.com/new-expensify/hubs/some-other-section/`

## Here is how I will achieve that

```js
        if (backButton) {
            backButton.addEventListener('click', function() {
                var currentPath = window.location.pathname;
                var pathSegments = currentPath.split('/');

                // Check if the URL structure is as expected
                if (pathSegments.length > 4 && pathSegments[2] === 'new-expensify') {
                    var section = pathSegments[3]; // This can be any section, like "getting-started"
                    var newPath = '/new-expensify/hubs/' + section + '/';
                    window.location.href = newPath;
                }
            });
        }
  
```

**Comment by rushatgabhane** [comment]
@Taiwrash why do we need listeners for this?

We know that the hub can be extracted from the URL, so can we change the `link` of the back button to the hub url? What do you think?

**Comment by Taiwrash** [comment]
@rushatgabhane,   You are right. If the `hub` can be directly extracted from the URL, it's more efficient to set the back button's link directly based on the URL. This way, we can avoid `event-listener` for performance efficiency.

## Check this implementation's out
```js
if (backLink) {
            var currentPath = window.location.pathname;
            var pathSegments = currentPath.split('/');

            // Check if the URL structure matches either "new-expensify" or "expensify-classic"
            if (pathSegments.length > 4 && (pathSegments[2] === 'new-expensify' || pathSegments[2] === 'expensify-classic')) {
                var product = pathSegments[2]; // This can be "new-expensify" or "expensify-classic"
                var section = pathSegments[3]; // This can be any section, like "getting-started"
                var newPath = '/' + product + '/hubs/' + section + '/';
                backLink.href = newPath;
            }
        }
```

**Comment by Taiwrash** [comment]
@rushatgabhane  @MitchExpensify any other comment here?

**Comment by Taiwrash** [comment]
Happy new week/month here...

Looking forward to a new update here...

**Comment by rushatgabhane** [comment]
@Taiwrash im not sure if we want to fix this.

**Comment by rushatgabhane** [comment]
@Taiwrash's proposal looks good

🎀👀🎀

**Comment by Taiwrash** [comment]
Is this not meant to be assigned to me?

The bot assigned to someone else....

- Can I go ahead to create the PR?

**Comment by Gonals** [comment]
> Is this not meant to be assigned to me?
> 
> The bot assigned to someone else....
> 
> * Can I go ahead to create the PR?

Hahaha, my job is to assign it to you 😁

**Comment by Taiwrash** [comment]
I just applied on upwork.

My PR will be available within 24 hours

**Comment by Taiwrash** [comment]
@rushatgabhane @MitchExpensify please when is the next review for this?

**Comment by MitchExpensify** [comment]
Reassigning while I'm on leave 🙇

**Comment by Taiwrash** [comment]
> Reassigning while I'm on leave 🙇

oh wish you a happy break!

**Comment by mallenexpensify** [comment]
Unsure of what's needed from me, looks like the PR is actively being reviewed so I'm gonna hang tight and follow along.

**Comment by Taiwrash** [comment]
> Unsure of what's needed from me, looks like the PR is actively being reviewed so I'm gonna hang tight and follow along.

Thanks @mallenexpensify. Don't forget to drop us your contribution that will speed up the process for us...

**Comment by Taiwrash** [comment]
**ok!** Currently, figuring out the bug!

**Comment by rushatgabhane** [comment]
@Taiwrash were you able to figure out the bug?

**Comment by Taiwrash** [comment]
> @Taiwrash were you able to figure out the bug?

I tried multiple attempts I have been able to pin-point the issue

**Comment by mallenexpensify** [comment]
@rushatgabhane what do you propose are best next steps here? Seems like we've stagnated a bit.

**Comment by Taiwrash** [comment]
I tried figuring it out but I am unable to fix the bugs pointed out.

**Comment by mallenexpensify** [comment]
Taking a stab at what I think is the best next steps
- Updated OP - When navigating on help.expensify.com, back should take you to the previous page
- Unassigned @Taiwrash 
- Added `Help Wanted`
- Made myself the issue owner (for now) so this doesn't gel lost in the abyss, also bumped to `Daily` for some reason. 

@rushatgabhane , if there are other contributor's who have experience with the help site, can you tag them here to see if anyone want to propose a solution?  Thx

**Comment by wildan-m** [comment]
@rushatgabhane is it open for new proposal?
can I confirm that the remaining issue is when go to section it's not accureate and create gap in the top of the section header?

<img width="981" alt="image" src="https://github.com/user-attachments/assets/0172c357-48c5-4411-b553-34fb36062b5c">

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

1. Back button on help.expensify.com should navigate to the previous page.
2. The LHN highlight position is inaccurate.

### What is the root cause of that problem?

1. We record the hashtag link in browser history each time we clicking the sidebar links
2. We add  `scrollSmoothOffset: -80` and headingsOffset: 80 in `tocbotOptions` for desktop, even desktop doesn't have fixed header

### What changes do you think we should make in order to solve the problem?

1. Don't add hash url in the first place when clicking the menu
add custom `onClick` in the tocbot options and `e.preventDefault()`, if we want to keep the hashText url, use history.replaceState to update the url

docs/assets/js/main.js
```tsx
    onClick: (e) => {
        e.preventDefault();
        const hashText = e.target.href.split('#').pop();
        // Append hashText to the current URL without saving to history
        const newUrl = `${window.location.pathname}#${hashText}`;
        history.replaceState(null, '', newUrl);
    },
```

2. Desktop doesn't have fixed header, use header and scrollling offset only for mobile, detect resolution and referesh the tocbot.
docs/assets/js/main.js 
```tsx
// Define the media query string for the mobile breakpoint
const mobileBreakpoint = window.matchMedia('(max-width: 799px)');

// Function to update tocbot options and refresh
function updateTocbotOptions(headingsOffset, scrollSmoothOffset) {
    tocbotOptions.headingsOffset = headingsOffset;
    tocbotOptions.scrollSmoothOffset = scrollSmoothOffset;
    window.tocbot.refresh({
        ...tocbotOptions,
        contentSelector: '#new-expensify',
    });
}

// Function to handle changes in the media query statuses
function handleBreakpointChange() {
    if (mobileBreakpoint.matches) {
        updateTocbotOptions(80, -80);
        console.log('[wildebug] Mobile breakpoint matched');
    } else {
        updateTocbotOptions(0, 0);
        console.log('[wildebug] Non-mobile breakpoint matched');
    }
}

// Add listener for changes to the media query status using addEventListener
mobileBreakpoint.addEventListener('change', handleBreakpointChange);

// Initial check
handleBreakpointChange();
```


[Branch for this solution.](https://github.com/Expensify/App/compare/main...wildan-m:wildan/fix/42600-fix-back-help-docs?expand=1)
### What alternative solutions did you explore? (Optional)
N/A

**Comment by wildan-m** [comment]
[Proposal Updated](https://github.com/Expensify/App/issues/42600#issuecomment-2303657933)

- remove unnecessary code

**Comment by Gonals** [comment]
Not overdue. We are getting some proposals

**Comment by mallenexpensify** [comment]
@rushatgabhane , 👀 on @wildan-m 's proposal above plz
https://github.com/Expensify/App/issues/42600#issuecomment-2303657933

**Comment by rushatgabhane** [comment]
@wildan-m 

`#` links are useful because they take you to a specific section on a page

I don't think we should remove them

**Comment by wildan-m** [comment]
@rushatgabhane PR ready for review at https://github.com/Expensify/App/pull/48273

**Comment by rushatgabhane** [comment]
@mallenexpensify PR in production 1 week ago. We can close here

**Comment by mallenexpensify** [comment]
Contributor: @wildan-m paid $250 via Upwork

@rushatgabhane , I'm assuming you're not paid here cuz you're working on help projects. 
hmm... do we want a regression test for this? Seems like we would, unless a test is being added elsewhere as part of a project.
